### ISSUE TITLE
False positive incompatible return type on decorator with TypeVar constraints

### ISSUE THREAD

**Issue by ziima** [python/mypy#14944]
**Bug Report**
I discovered a difference in cases where I create a decorator for async function depending on whether I use `bound` or `constraints` arguments of `TypeVar`.

```python
from typing import Any, Awaitable, Callable, TypeVar, cast


Func = TypeVar('Func', bound=Callable[[int], Awaitable[int]])


def my_dec(func: Func) -> Func:
    async def wrapper(value: Any) -> int:
        reply = await func(value)
        return reply
    return cast(Func, wrapper)


Func2 = TypeVar('Func2', Callable[[int], Awaitable[int]], Callable[[str], Awaitable[int]])


def my_dec2(func: Func2) -> Func2:
    async def wrapper(value: Any) -> int:
        reply = await func(value)
        return reply  # error: Incompatible return value type (got "int", expected "Coroutine[Any, Any, int]")  [return-value]
    return cast(Func2, wrapper)
```

**To Reproduce**
https://mypy-play.net/?mypy=latest&python=3.11&gist=000d7034a44109d673d7b12be9a46ae0

**Expected Behavior**

Mypy should infer that `reply` type `int` is valid and shouldn't expect a `Coroutine` object.

**Actual Behavior**

See comment in the snippet.

**Your Environment**

- Mypy version used: 1.1.1 (multiple)
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.11 (multiple)

**Comment by jttimko5** [comment]
Hi @ziima, Id like to work on this if thats okay, can you please assign it to me?

**Comment by ziima** [comment]
@jttimko5 I'm not a maintainer, so I can't.

**Comment by Taywee** [comment]
I ran into this as well, and opened issue #14979 before I found this one.  I'll put my simple reproductions here, because it's actually incredibly easy to trigger this issue.  You don't need to await or even use the passed-in argument:

[Here's an example in 5 lines with one import that does it.](https://mypy-play.net/?mypy=latest&python=3.11&gist=90b075b2d981c70148162640c1621222)

**Comment by jttimko5** [comment]
Could someone provide some guidance about where I would look to implement this feature? I am new to the codebase so any help pointing me in the right direction would be helpful. Thanks!

**Comment by erictraut** [comment]
It appears this was fixed in mypy 1.4.0. It no longer repros for me.
